docs/proposals/20221109-managed-external-etcd.md

```diff
@@ -0,0 +1,572 @@
+- Provide a first implementation of this pluggable etcd provider using [etcdadm](https://github.com/kubernetes-sigs/etcdadm) that integrates with the Kubeadm Control Plane provider.
+  - The etcdadm based provider will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+  - Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
+
+
+### Non-Goals/Future work
+- The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available. Replace IP addresses/hostnames with static endpoints once the Load Balancer provider is available, or explore using kube-vip for this use case.
+- Modifying the control plane provider contract. The first pass will introduce an external etcd provider that only works with the KubeadmControlPlane provider. A separate proposal would include some API changes to the control plane contract to make this work with any control plane provider.
+- Adding a new API type to represent etcd Machine object. This first iteration will reuse the existing Machine object to represent an etcd machine.
+- This is not to be used for etcd-as-a-service. The etcd clusters created by an etcd provider within CAPI will only be used by the kube-apiserver of a target workload cluster.
+- Etcd cluster snapshots will need its own separate KEP. It should include details about how to take a snapshot, if `etcdctl snapshot save` command is used then how to collect the snapshots from the etcd machines and an integration with a storage provider to save those snapshots. This first iteration will not have a snapshot mechanism implemented. Users can still take snapshots by manually running the `etcdctl snapshot save` command and we can document those steps.
+
+## Proposal
+
+### User Stories
+
+- As an end user, I want to be able to create a Kubernetes cluster that uses external etcd topology using CAPI with a single step.
+- On creating the required CRs in CAPI along with the new etcd provider specific CRs, the external etcd provider should provision the etcd cluster first, and provide that as input to the workload cluster, without requiring any intervention from me as an end user.
+- As an end user, I should be able to use the etcd provider CRs to specify etcd cluster size during creation and specify/modify etcd version. CAPI controllers should modify and manage the etcd clusters accordingly.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model Changes
+
+The following new type/CRDs will be added for the etcdadm based provider
+```go
+// API Group: etcd.cluster.x-k8s.io OR etcdplane.cluster.x-k8s.io
+type EtcdadmCluster struct {
```

Comment:
Could we avoid relying on `etcdadm` here, and just call it `etcd`? Reason being, if in the future we'd like to swap out the implementation detail, we should be able to do so

Reply:
Oh that's an interesting point. However, wouldn't that almost guarantee another provider? Similarly to if we ever want to swap kubeadm with a different tool?

Reply:
It might be better to replace `etcdadm based provider` with `etcdadm based etcd provider` to avoid confusion. IMO calling `etcdadm` here would be fine because that's one of the reference implementations of the etcd provider.